sql/src/main/java/org/apache/druid/sql/calcite/external/TableAppendMacro.java

```diff
@@ -0,0 +1,344 @@
+          }
+        }
+      }
+      return true;
+    }
+
+    @Override
+    public SqlOperandCountRange getOperandCountRange()
+    {
+      return SqlOperandCountRanges.from(1);
+    }
+
+    @Override
+    public String getAllowedSignatures(SqlOperator op, String opName)
+    {
+      return "APPEND( <TABLE_NAME>[, <TABLE_NAME> ...] )";
+    }
+
+    @Override
+    public List<RelDataType> paramTypes(RelDataTypeFactory typeFactory)
+    {
+      RelDataType t = typeFactory.createSqlType(SqlTypeName.VARCHAR);
+      return ImmutableList.<RelDataType>builder().add(t, t).build();
+    }
+
+    @Override
+    public List<String> paramNames()
+    {
+      return ImmutableList.<String>builder().add("tableName", "tableName").build();
+    }
+  }
+
+  @Override
+  public List<String> getParamNames()
+  {
+    return ImmutableList.<String>builder().add("tableName", "tableName").build();
+  }
+
+  @Override
+  public TranslatableTable getTable(SqlOperatorBinding operatorBinding)
+  {
+    SqlCallBinding callBinding = (SqlCallBinding) operatorBinding;
+    SqlValidator validator = callBinding.getValidator();
+
+    List<TableOperand> tables = getTables(callBinding, validator.getCatalogReader());
+    AppendDatasourceMetadata metadata = buildUnionDataSource(tables);
+    return new DatasourceTable(
+        metadata.values,
+        metadata,
+        EffectiveMetadata.of(metadata.values)
+    );
+  }
+
+  static class TableOperand
+  {
+    private final SqlNode sqlOperand;
+    private final SqlValidatorTable table;
+
+    public TableOperand(SqlNode sqlOperand, SqlValidatorTable table)
+    {
+      this.sqlOperand = sqlOperand;
+      this.table = table;
+    }
+
+    public RelOptTable getRelOptTable()
+    {
+      return table.unwrapOrThrow(RelOptTable.class);
+    }
+
+    public DatasourceTable getDataSourceTable()
+    {
+      return table.unwrapOrThrow(DatasourceTable.class);
+    }
+
+    public RowSignature getRowSignature()
+    {
+      return getDataSourceTable().getRowSignature();
+    }
+
+    public DataSource getDataSource()
+    {
+      return getDataSourceTable().getDataSource();
+    }
+  }
+
+  private List<TableOperand> getTables(SqlCallBinding callBinding, SqlValidatorCatalogReader catalogReader)
+  {
+    List<TableOperand> tables = new ArrayList<>();
+    for (int i = 0; i < callBinding.getOperandCount(); i++) {
+      if (!callBinding.isOperandLiteral(i, false)) {
+        throw DruidSqlValidator.buildCalciteContextException(
+            "All arguments of call to macro "
+                + "APPEND should be literal. Actual argument #"
+                + i + " is not literal",
+                callBinding.operand(i)
+        );
+      }
+      @Nullable
+      String tableName = callBinding.getOperandLiteralValue(i, String.class);
+      ImmutableList<String> tableNameList = ImmutableList.<String>builder().add(tableName).build();
+      SqlValidatorTable table = catalogReader.getTable(tableNameList);
```

Comment:
does this also take care of tables created via CTEs?

Reply:
Thanks for the explanation - it makes sense to only support persisted tables for now. This can be considered when the execution engine supports CTE + table union easily. 
We can mention this caveat in the documentation for the users though.

Reply:
added the following `info` to the docs:

> [!NOTE]
> Only tables defined in the catalog are supported in `TABLE(APPEND())` - due to that; common table expressions result in `table not found` errors for queries like:
> ```sql
> WITH cte_table AS (SELECT * from TABLE1) SELECT * FROM TABLE(APPEND('cte_table'))
> ```